### PR TITLE
Add scheduleWarmUpFrame

### PR DESCRIPTION
**Pull request by dkwingsmt**
This PR adds `PlatformDispatcher.scheduleWarmUpFrame`.

This PR is needed for the follow up changes:
* The framework will switch to using this function to render warmup frames in https://github.com/flutter/flutter/pull/143290.
* Then the engine will finally be able to switch to multiview pipeline with no regression on startup timing in https://github.com/flutter/engine/pull/49950.

For why the warm up frame must involve the engine to render, see https://github.com/flutter/flutter/issues/142851.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: lib/ui/platform_dispatcher.dart
Comment (loic-sharma) [inline lib/ui/platform_dispatcher.dart:805]: I'm concerned folks might incorrectly use `forceSyncFrame` here thinking it might improve performance. Perhaps we shouldn't advertise `forceSyncFrame` at all here?

File: lib/ui/platform_dispatcher.dart
Comment (loic-sharma) [inline lib/ui/platform_dispatcher.dart:828]: This can cause screen tearing & web doesn't support "out-of-vsync rendering".

Perhaps:

```suggestion
  /// Immediately render a frame by invoking the [onBeginFrame] and
  /// [onDrawFrame] callbacks synchronously.
  ///
  /// This forces a frame to be computed synchronously within this call,
  /// ignoring v-sync. Rendering in a forced frame may not be fully
  /// supported by all platforms.
  ///
  /// Prefer [scheduleFrame] to update the display in normal operation. The
  /// [forceSyncFrame] method is designed for situations that require a frame is
  /// rendered as soon as possible, even at the cost of rendering quality. An
  /// example of using this method is [SchedulerBinding.scheduleWarmUpFrame],
  /// which is called during application startup so that resources can be prepared
  /// before the first frame.
  ///
  /// See also:
  ///
  ///  * [SchedulerBinding.scheduleWarmUpFrame], which uses this method.
  ///  * [scheduleFrame].
```

File: lib/web_ui/lib/src/engine/initialization.dart
Comment (loic-sharma) [inline lib/web_ui/lib/src/engine/initialization.dart:201]: Consider making this a named argument for readability

File: shell/common/animator.h
Comment (loic-sharma) [inline shell/common/animator.h:56]: Please add a quick comment here

File: shell/common/animator.cc
Comment (loic-sharma) [inline shell/common/animator.cc:278]: `BeginFrame` won't call the begin frame / draw frame callbacks if the frame pipeline is full. I think that's the desired behavior, could you confirm? If yes, we should call this out on the `dart:ui` docs.

File: lib/ui/platform_dispatcher.dart
Comment (dnfield) [reply lib/ui/platform_dispatcher.dart:805]: We should not create an API like this.

People _will_ use it incorrectly and misunderstand what it is for.

This should be named as it is in the current framework usage, something like `requestWarmupFrame`.

File: lib/ui/platform_dispatcher.dart
Comment (dnfield) [inline lib/ui/platform_dispatcher.dart:812]: We should not document that this happens synchronously. We should just document that this allows initiation of a warmup frame independent of whether the engine has processed a vsync signal yet, and that it is intended for use only for the purpose of rendering a warmup frame at program startup.

**Review by dnfield — CHANGES_REQUESTED** [review]

**Comment by dnfield** [comment]
One alternative that may be even simpler is to encode the way the framework works here into the engine itself.

In other words, rather than allowing the framework to request this, we just fire off the first request for rendering _ASAP_ instead of waiting for the vsync handler to be hooked up.

That may avoid some of the confusion around how to use this API. At any rate, we should not create an API that makes application authors think they can call this whenever they want a "synchronous frame" to render.

**Comment by dkwingsmt** [comment]
> One alternative that may be even simpler is to encode the way the framework works here into the engine itself.
> 
> In other words, rather than allowing the framework to request this, we just fire off the first request for rendering _ASAP_ instead of waiting for the vsync handler to be hooked up.
> 
> That may avoid some of the confusion around how to use this API. At any rate, we should not create an API that makes application authors think they can call this whenever they want a "synchronous frame" to render.

Yeah I agree. During my discussion with @/goderbauer and @/loic-sharma we also wanted to make the name sound intimidating for people to not use it lightly. 

I like your idea, but it's not very feasible since this requires that the app (not necessarily the framework; the app could use dart:ui directly) assigning the warm up frame callback, and that the engine knows when it's assigned. 

Since the objection is strong, I'll change the API name.

**Comment by jonahwilliams** [comment]
Specifically, I think the name should include "warmup" in the name

**Comment by dkwingsmt** [comment]
I've overhauled this PR by adopting the two-callback approach suggested by @dnfield. This PR should be ready for review except for some unit tests for the new API.
@jonahwilliams @loic-sharma @goderbauer 

The accompanying PRs have also be updated as well, for reference:
* https://github.com/flutter/flutter/pull/143290, which should be merged after this PR and make the framework use the new API.
* https://github.com/flutter/engine/pull/49950, which should be merged after the framework PR and make the pipeline multi-view.

File: shell/common/animator.cc
Comment (dkwingsmt) [reply shell/common/animator.cc]: In the new doc, I've explicitly said that the frame might or might not be rendered depending on situations and platforms.

**Review by loic-sharma — APPROVED** [review]
LGTM but please also get approvals from @dnfield, @jonahwilliams, and someone on the web team before merging this

File: shell/common/animator.cc
Comment (goderbauer) [inline shell/common/animator.cc:268]:
```suggestion
  // Do nothing. The warm up frame does not need any additional work to end the
```

I think?

File: lib/web_ui/lib/src/engine/platform_dispatcher.dart
Comment (goderbauer) [inline lib/web_ui/lib/src/engine/platform_dispatcher.dart:781]: Does this TODO make sense here? My understanding is that because of the `Timer.run`s the warmUp frame actually is (and always has been) flushing microtasks in between?

That's also what the comment in the other file seems to imply:

> // We use timers here to ensure that microtasks flush in between.

**Review by goderbauer — COMMENTED** [review]
Can some test coverage be added for the new dart:ui API?

File: lib/web_ui/lib/src/engine/platform_dispatcher.dart
Comment (dkwingsmt) [reply lib/web_ui/lib/src/engine/platform_dispatcher.dart:781]: You're right! Thank you very much for pointing it out. It made sense in the old code but I forgot to rethink it for the new code.

File: lib/ui/platform_dispatcher.dart
Comment (dnfield) [inline lib/ui/platform_dispatcher.dart:832]: nit: Why not `FrameCallback` for beginFrame?

**Review by dnfield — COMMENTED** [review]
Over all looks good.

Please add a test under //flutter/testing/dart/platform_dispatcher_test.dart.

Consider, for example, a test that asserts that a microtask scheduled in the `beginFrame` callback has executed by the time the `drawFrame` callback is run.

File: lib/ui/platform_dispatcher.dart
Comment (dkwingsmt) [reply lib/ui/platform_dispatcher.dart:832]: In the warm up frame, the frame time is actually not used. https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/scheduler/binding.dart#L1031

(To dig deeper, the frame time of the first beginFrame is used to set the epoch. It's specially intended for the warm up frame to not set the epoch.)

**Comment by dkwingsmt** [comment]
@dnfield Thanks for the review. I've added a test.

The test does not verify the microtask execution. I wanted to, but the microtasks won't run. I've instead added this test to the framework (https://github.com/flutter/flutter/pull/143290/files#diff-bc524bd4a8c5b61d9c667ce8aa71f07c585559899471a638c627040247727604R134) and it works. Apparently the framework test uses a different setup from just running the engine. (By close examination, the framework test uses a custom `Zone`.) I'm not sure yet if a real app uses the same setup as the framework test.

File: shell/common/fixtures/shell_test.dart
Comment (dnfield) [inline shell/common/fixtures/shell_test.dart:549]: This seems like a strange test to me - this API is documented as saying not to call it at arbitrary times, and metrics changes can happen for all sorts of reasons.

If anything, I'd expect we make this test fail at some point rather than continue to support it, right?

**Comment by dnfield** [comment]
Here's the testI had in mind - it passes for me with your patch

```dart
test('PlatformDispatcher.scheduleWarmupFrame', () async {
    bool microtaskFlushed = false;
    bool beginFrameCalled = false;
    final Completer<void> drawFrameDone = Completer<void>();
    PlatformDispatcher.instance.scheduleWarmUpFrame(beginFrame: () {
      print('beginFrame');
      expect(microtaskFlushed, false);
      expect(drawFrameDone.isCompleted, false);
      expect(beginFrameCalled, false);
      beginFrameCalled = true;
      scheduleMicrotask(() {
        print('microtask!');
        expect(microtaskFlushed, false);
        expect(drawFrameDone.isCompleted, false);
        microtaskFlushed = true;
      });
      expect(microtaskFlushed, false);
    }, drawFrame: () {
      print('drawFrame');
      expect(beginFrameCalled, true);
      expect(microtaskFlushed, true);
      expect(drawFrameDone.isCompleted, false);
      drawFrameDone.complete();
    });
    await drawFrameDone.future;
    expect(beginFrameCalled, true);
    expect(drawFrameDone.isCompleted, true);
    expect(microtaskFlushed, true);
  });
```

File: shell/common/fixtures/shell_test.dart
Comment (dkwingsmt) [reply shell/common/fixtures/shell_test.dart]: I've managed to simplify the test. Now the dart program simply schedules the warm up frame immediately, since I realized that the frame can be scheduled from the engine side to create the continuation.

**Comment by dkwingsmt** [comment]
@dnfield Thank you! I've added your test to the PR.

**Review by dnfield — APPROVED** [review]

**Review by jonahwilliams — APPROVED** [review]
LGTM

File: lib/ui/platform_dispatcher.dart
Comment (yjbanov) [inline lib/ui/platform_dispatcher.dart:814]: This phrasing is unclear from the standpoint of an author of an app or framework that does not use the standard Flutter framework (e.g. a game engine, or a framework with a different design). What does "is used" mean? Used by what?

File: lib/ui/platform_dispatcher.dart
Comment (yjbanov) [inline lib/ui/platform_dispatcher.dart:832]: If this is meant for first frame only, why not call it `scheduleFirstFrame`?

File: lib/ui/platform_dispatcher.dart
Comment (yjbanov) [inline lib/ui/platform_dispatcher.dart:818]: Do we also want to use this when rendering the first frame in a new `FlutterView`? I imagine we'd want the new view to be rendered as fast as possible too.

File: lib/web_ui/lib/src/engine/platform_dispatcher.dart
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/platform_dispatcher.dart:789]: I'm not sure timers is the right choice. Timers have variable precision on the web to protect users from timing attacks. There's no guarantee that a timer will fire before the next vsync. I think we should call `beginFrame` and `drawFrame` back to back synchronously in a microtask. We can flush microtasks using a `Zone`.

However, in the web case, I think there's not much difference between a "warm-up" frame and a regular `requestAnimationFrame`. The actual first frame is rendered by the HTML engine. Flutter is only ever renders subsequent frames. So perhaps on the web this should still be a `requestAnimationFrame`.

Comment (yjbanov) [reply lib/web_ui/lib/src/engine/platform_dispatcher.dart:789]: Actually, we don't flush microtasks between the two on the web anyway: https://github.com/flutter/engine/blob/23d34eaa5007d185eb3edcd97c85d387ded2a665/lib/web_ui/lib/src/engine/initialization.dart#L181

So I'd just call them back-to-back in a microtask, something like this:

```dart
scheduleMicrotask(() {
  beginFrame();
  drawFrame();
});
```

Comment (dkwingsmt) [reply lib/web_ui/lib/src/engine/platform_dispatcher.dart:789]: This comment is moved from the framework. https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/scheduler/binding.dart#L1028

I agree that Web might want to implement this method differently, especially since Web doesn't even flush the microevents during a regular frame. I'll change it to the following block, is it ok?
```
          beginFrame();
          // TODO(yjbanov): technically Flutter flushes microtasks between
          //                onBeginFrame and onDrawFrame. We don't, which hasn't
          //                been an issue yet, but eventually we'll have to
          //                implement it properly. (Same TODO as in
          //                `initializeEngineServices`).
          endFrame();
```

However, I think keeping the "warm up" behavior (and hence name) here is important. The warm up frame (as documented in [`SchedulerBinding.scheduleWarmUpFrame`](https://github.com/flutter/flutter/blob/f9b3b84d4eb5afdbe9159fb5ac7a7d24a939a3e6/packages/flutter/lib/src/scheduler/binding.dart#L996)) is designed to make the framework build the first layer tree (cold start) sooner. Therefore this is better done as soon as possible instead of in the next vsync/`requestAnimationFrame`. It's ok if the frame is not rendered at all on the Web.

File: lib/ui/platform_dispatcher.dart
Comment (dkwingsmt) [reply lib/ui/platform_dispatcher.dart:832]: The biggest reason, in my opinion, is that this frame is not guaranteed to be rendered, for example if `Animator::BeginFrame` is not called before, or on Web (if you agree with [my other comment](https://github.com/flutter/engine/pull/50570#discussion_r1496722312)). Therefore this might not be the first frame.

Besides, I don't want to give developers the impression that scheduling the first frame using `scheduleFrame` is wrong. The `scheduleWarmUpFrame` really is for warming up. An app doesn't need warm-ups to perform correctly, but will perform better.

File: lib/ui/platform_dispatcher.dart
Comment (dkwingsmt) [reply lib/ui/platform_dispatcher.dart:814]: This paragraph is copied from `SchedulerBinding.scheduleWarmUpFrame`, but I agree with your concern. Do you think the following way would be better?

```
  /// The application can call this method as soon as it starts up so that the first
  /// frame (which is likely to be quite expensive) can start a few extra
  /// milliseconds earlier. [...]
```

File: lib/ui/platform_dispatcher.dart
Comment (dkwingsmt) [reply lib/ui/platform_dispatcher.dart:818]: That's a good point, although I imagine this would require more work. Can I open an issue to leave the optimization for later?

File: lib/web_ui/lib/src/engine/platform_dispatcher.dart
Comment (dkwingsmt) [reply lib/web_ui/lib/src/engine/platform_dispatcher.dart:789]: (Sorry I made the comment without seeing your reply) Sure! I'll do it. Should I keep the TODO though?

Comment (dkwingsmt) [reply lib/web_ui/lib/src/engine/platform_dispatcher.dart:789]: Also, should it be `scheduleMicroTask` or `Timer.run` here? According to the doc of `Timer.run`, 
```
  /// Runs the given [callback] asynchronously as soon as possible.
```
I think `Timer.run` would be the better choice.

File: lib/ui/platform_dispatcher.dart
Comment (goderbauer) [reply lib/ui/platform_dispatcher.dart:833]: In addition to what Tong said: The warmup frame is also used in hot reload scenarios where it isn't the first frame of the app.

File: lib/web_ui/lib/src/engine/platform_dispatcher.dart
Comment (goderbauer) [reply lib/web_ui/lib/src/engine/platform_dispatcher.dart]: Today, the framework handles warmup-scheduling and it uses timers here even on the web. This change just moves that scheduling logic to dart:ui. Since the bootstrapping logic is somewhat brittle, I would argue we should keep the existing behavior here for now and use Timers for warn-up frames. If we want a change in behavior here for the web and do something else I would suggest doing that in a separate follow-up PR to decouple that from the other changes in case something goes wrong.

Just my 50c, ultimately I leave it up to the web team to make a call.

File: lib/ui/platform_dispatcher.dart
Comment (goderbauer) [reply lib/ui/platform_dispatcher.dart]: I don't think this optimization works for individual views. Because of the unified widget tree we can only produce frames for all views at once and cannot target individual views. This optimization only works for the app as a whole.

File: lib/web_ui/test/engine/platform_dispatcher/platform_dispatcher_test.dart
Comment (goderbauer) [inline lib/web_ui/test/engine/platform_dispatcher/platform_dispatcher_test.dart:421]: nit: remove one of the blank lines?

File: lib/web_ui/lib/src/engine/platform_dispatcher.dart
Comment (jonahwilliams) [reply lib/web_ui/lib/src/engine/platform_dispatcher.dart]: I agree with @goderbauer , lets not combine refactors and logic changes into the same PR.

Comment (yjbanov) [reply lib/web_ui/lib/src/engine/platform_dispatcher.dart:795]: Ah, I didn't know we already did this. In that case I'm assuming that the framework already resolves the race condition that I was worried about (e.g. it will not schedule the second frame until the first frame is done). Yes, I agree let's keep the timers for now.

File: lib/ui/platform_dispatcher.dart
Comment (yjbanov) [reply lib/ui/platform_dispatcher.dart]: Yep, SGTM!

File: lib/ui/platform_dispatcher.dart
Comment (yjbanov) [reply lib/ui/platform_dispatcher.dart:833]: The semantics of this method sound pretty complicated. I guess warm up is fine then. Should we have some assertions in the engine for this method? For example, should we make sure it's only called for the first frame, and is never called after a `scheduleFrame` is called? Simply exposing this as a sibling to `scheduleFrame` seems to contain footguns.

File: lib/ui/platform_dispatcher.dart
Comment (yjbanov) [reply lib/ui/platform_dispatcher.dart]: Nope, I was just hypothesizing. If we don't need it, then we don't need it. My gut feeling is that we don't need any of this on the web. So I'm actually a little sad that this API even exists.

**Review by yjbanov — APPROVED** [review]
LGTM. None of my comments are blocking.

**Pull request merged**
